// ts-sql-query/queryBuilders/InsertQueryBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InsertQueryBuilder = void 0;
const SqlBuilder_1 = require("../sqlBuilders/SqlBuilder");
const ITableOrView_1 = require("../utils/ITableOrView");
const Column_1 = require("../utils/Column");
const chained_error_1 = require("chained-error");
const attachSource_1 = require("../utils/attachSource");
const symbols_1 = require("../utils/symbols");
const values_1 = require("../expressions/values");
const ITableOrView_2 = require("../utils/ITableOrView");
const ComposeSliptQueryBuilder_1 = require("./ComposeSliptQueryBuilder");
// one implement ommited intentionally to don't confuse TypeScript
class InsertQueryBuilder extends ComposeSliptQueryBuilder_1.ComposeSplitQueryBuilder {
    constructor(sqlBuilder, table) {
        super(sqlBuilder);
        this.__sets = {};
        this.__isMultiple = false;
        this.__withs = [];
        // cache
        this.__query = '';
        this.__params = [];
        this.__table = table;
        (0, ITableOrView_1.__getTableOrViewPrivate)(table).__addWiths(this.__sqlBuilder, this.__withs);
    }
    executeInsert(min, max) {
        this.query();
        const source = new Error('Query executed at');
        try {
            const idColumn = this.__idColumn;
            const multiple = this.__multiple;
            let result;
            let returningLastInsertedId = !idColumn;
            if (multiple && multiple.length <= 0) {
                if (idColumn) {
                    return this.__sqlBuilder._queryRunner.createResolvedPromise([]);
                }
                else {
                    return this.__sqlBuilder._queryRunner.createResolvedPromise(0);
                }
            }
            else if (!idColumn) {
                result = this.__sqlBuilder._queryRunner.executeInsert(this.__query, this.__params).catch((e) => {
                    throw (0, attachSource_1.attachSource)(new chained_error_1.default(e), source);
                });
            }
            else if (!multiple && !this.__from) {
                result = this.__sqlBuilder._queryRunner.executeInsertReturningLastInsertedId(this.__query, this.__params).then((value) => {
                    if (value === undefined) {
                        value = null;
                    }
                    const idColumnPrivate = (0, Column_1.__getColumnPrivate)(idColumn);
                    const typeAdapter = idColumnPrivate.__typeAdapter;
                    let result;
                    if (typeAdapter) {
                        result = typeAdapter.transformValueFromDB(value, idColumnPrivate.__valueType, this.__sqlBuilder._defaultTypeAdapter);
                    }
                    else {
                        result = this.__sqlBuilder._defaultTypeAdapter.transformValueFromDB(value, idColumnPrivate.__valueType);
                    }
                    if (!this.onConflictDoNothing && (result === null || result === undefined)) {
                        throw new Error('Expected a value as result of the insert returning last inserted id, but null or undefined value was found');
                    }
                    if (this.__isMultiple) {
                        return [result];
                    }
                    else {
                        return result;
                    }
                }).catch((e) => {
                    throw (0, attachSource_1.attachSource)(new chained_error_1.default(e), source);
                });
            }
            else {
                result = this.__sqlBuilder._queryRunner.executeInsertReturningMultipleLastInsertedId(this.__query, this.__params).then((rows) => {
                    const idColumnPrivate = (0, Column_1.__getColumnPrivate)(idColumn);
                    const typeAdapter = idColumnPrivate.__typeAdapter;
                    const columnType = idColumnPrivate.__valueType;
                    const defaultTypeAdapter = this.__sqlBuilder._defaultTypeAdapter;
                    if (typeAdapter) {
                        return rows.map((row, index) => {
                            const result = typeAdapter.transformValueFromDB(row, columnType, defaultTypeAdapter);
                            if (result === null || result === undefined) {
                                throw new Error('Expected a value as result of the insert returning last inserted id, but null or undefined value was found at index ' + index);
                            }
                            return result;
                        });
                    }
                    else {
                        return rows.map((row, index) => {
                            const result = defaultTypeAdapter.transformValueFromDB(row, columnType);
                            if (result === null || result === undefined) {
                                throw new Error('Expected a value as result of the insert returning last inserted id, but null or undefined value was found at index ' + index);
                            }
                            return result;
                        });
                    }
                }).catch((e) => {
                    throw (0, attachSource_1.attachSource)(new chained_error_1.default(e), source);
                });
            }
            if (min !== undefined) {
                result = result.then((result) => {
                    let count;
                    if (Array.isArray(result)) {
                        count = result.length;
                    }
                    else if (returningLastInsertedId) {
                        if (result === null || result === undefined) {
                            count = 0;
                        }
                        else {
                            count = 1;
                        }
                    }
                    else {
                        count = result;
                    }
                    if (count < min) {
                        throw (0, attachSource_1.attachSource)(new Error("The insert operation didn't insert the minimum of " + min + " row(s)"), source);
                    }
                    if (max !== undefined && count > max) {
                        throw (0, attachSource_1.attachSource)(new Error("The insert operation insert more that the maximum of " + max + " row(s)"), source);
                    }
                    return result;
                });
            }
            return result;
        }
        catch (e) {
            throw new chained_error_1.default(e);
        }
    }
    executeInsertNoneOrOne() {
        this.query();
        const source = new Error('Query executed at');
        try {
            this.__sqlBuilder._resetUnique();
            let result;
            if (this.__oneColumn) {
                result = this.__sqlBuilder._queryRunner.executeInsertReturningOneColumnOneRow(this.__query, this.__params).then((value) => {
                    const valueSource = this.__columns['result'];
                    if (!(0, values_1.isValueSource)(valueSource)) {
                        throw new Error('The result column must be a ValueSource');
                    }
                    if (value === undefined) {
                        return null;
                    }
                    return this.__transformValueFromDB(valueSource, value);
                }).catch((e) => {
                    throw (0, attachSource_1.attachSource)(new chained_error_1.default(e), source);
                });
            }
            else {
                result = this.__sqlBuilder._queryRunner.executeInsertReturningOneRow(this.__query, this.__params).then((row) => {
                    if (row) {
                        return this.__transformRow(row);
                    }
                    else {
                        return null;
                    }
                }).catch((e) => {
                    throw (0, attachSource_1.attachSource)(new chained_error_1.default(e), source);
                });
            }
            return this.__applyCompositions(result, source);
        }
        catch (e) {
            throw new chained_error_1.default(e);
        }
    }
    executeInsertOne() {
        this.query();
        const source = new Error('Query executed at');
        try {
            this.__sqlBuilder._resetUnique();
            let result;
            if (this.__oneColumn) {
                result = this.__sqlBuilder._queryRunner.executeInsertReturningOneColumnOneRow(this.__query, this.__params).then((value) => {
                    const valueSource = this.__columns['result'];
                    if (!(0, values_1.isValueSource)(valueSource)) {
                        throw new Error('The result column must be a ValueSource');
                    }
                    if (value === undefined) {
                        throw new Error('No result returned by the database');
                    }
                    return this.__transformValueFromDB(valueSource, value);
                }).catch((e) => {
                    throw (0, attachSource_1.attachSource)(new chained_error_1.default(e), source);
                });
            }
            else {
                result = this.__sqlBuilder._queryRunner.executeInsertReturningOneRow(this.__query, this.__params).then((row) => {
                    if (row) {
                        return this.__transformRow(row);
                    }
                    else {
                        throw new Error('No result returned by the database');
                    }
                }).catch((e) => {
                    throw (0, attachSource_1.attachSource)(new chained_error_1.default(e), source);
                });
            }
            return this.__applyCompositions(result, source);
        }
        catch (e) {
            throw new chained_error_1.default(e);
        }
    }
    executeInsertMany(min, max) {
        const source = new Error('Query executed at');
        this.query();
        try {
            this.__sqlBuilder._resetUnique();
            let result;
            if (this.__oneColumn) {
                result = this.__sqlBuilder._queryRunner.executeInsertReturningOneColumnManyRows(this.__query, this.__params).then((values) => {
                    const valueSource = this.__columns['result'];
                    if (!(0, values_1.isValueSource)(valueSource)) {
                        throw new Error('The result column must be a ValueSource');
                    }
                    return values.map((value) => {
                        if (value === undefined) {
                            value = null;
                        }
                        return this.__transformValueFromDB(valueSource, value);
                    });
                }).catch((e) => {
                    throw (0, attachSource_1.attachSource)(new chained_error_1.default(e), source);
                });
            }
            else {
                result = this.__sqlBuilder._queryRunner.executeInsertReturningManyRows(this.__query, this.__params).then((rows) => {
                    return rows.map((row, index) => {
                        return this.__transformRow(row, index);
                    });
                }).catch((e) => {
                    throw (0, attachSource_1.attachSource)(new chained_error_1.default(e), source);
                });
            }
            if (min !== undefined) {
                result = result.then((rows) => {
                    const count = rows.length;
                    if (count < min) {
                        throw (0, attachSource_1.attachSource)(new Error("The insert operation didn't insert the minimum of " + min + " row(s)"), source);
                    }
                    if (max !== undefined && count > max) {
                        throw (0, attachSource_1.attachSource)(new Error("The insert operation insert more that the maximum of " + max + " row(s)"), source);
                    }
                    return rows;
                });
            }
            return this.__applyCompositions(result, source);
        }
        catch (e) {
            throw new chained_error_1.default(e);
        }
    }
    query() {
        if (this.__query) {
            return this.__query;
        }
        try {
            if (this.__from) {
                this.__query = this.__sqlBuilder._buildInsertFromSelect(this, this.__params);
            }
            else if (this.__multiple) {
                this.__query = this.__sqlBuilder._buildInsertMultiple(this, this.__params);
            }
            else if (this.__sets === DEFAULT_VALUES) {
                this.__query = this.__sqlBuilder._buildInsertDefaultValues(this, this.__params);
            }
            else {
                this.__query = this.__sqlBuilder._buildInsert(this, this.__params);
            }
        }
        catch (e) {
            throw new chained_error_1.default(e);
        }
        return this.__query;
    }
    params() {
        if (!this.__query) {
            this.query();
        }
        return this.__params;
    }
    __toSql(_sqlBuilder, params) {
        if (this.__from) {
            return this.__sqlBuilder._buildInsertFromSelect(this, params);
        }
        else if (this.__multiple) {
            return this.__sqlBuilder._buildInsertMultiple(this, params);
        }
        else if (this.__sets === DEFAULT_VALUES) {
            return this.__sqlBuilder._buildInsertDefaultValues(this, params);
        }
        else {
            return this.__sqlBuilder._buildInsert(this, params);
        }
    }
    __toSqlForCondition(sqlBuilder, params) {
        return this.__toSql(sqlBuilder, params);
    }
    dynamicSet() {
        this.__query = '';
        return this;
    }
    set(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            const value = columns[property];
            sets[property] = value;
        }
        return this;
    }
    setIfValue(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            const value = columns[property];
            if (!this.__sqlBuilder._isValue(value)) {
                continue;
            }
            sets[property] = value;
        }
        return this;
    }
    setIfSet(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            if (!(property in sets)) {
                continue;
            }
            const value = columns[property];
            sets[property] = value;
        }
        return this;
    }
    setIfSetIfValue(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            if (!(property in sets)) {
                continue;
            }
            const value = columns[property];
            if (!this.__sqlBuilder._isValue(value)) {
                continue;
            }
            sets[property] = value;
        }
        return this;
    }
    setIfNotSet(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            if (property in sets) {
                continue;
            }
            const value = columns[property];
            sets[property] = value;
        }
        return this;
    }
    setIfNotSetIfValue(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            if (property in sets) {
                continue;
            }
            const value = columns[property];
            if (!this.__sqlBuilder._isValue(value)) {
                continue;
            }
            sets[property] = value;
        }
        return this;
    }
    ignoreIfSet(...columns) {
        this.__query = '';
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
        }
        else {
            sets = this.__sets;
        }
        for (let i = 0, length = columns.length; i < length; i++) {
            let column = columns[i];
            delete sets[column];
        }
        return this;
    }
    setIfHasValue(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            if (!this.__sqlBuilder._isValue(sets[property])) {
                continue;
            }
            const value = columns[property];
            sets[property] = value;
        }
        return this;
    }
    setIfHasValueIfValue(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            if (!this.__sqlBuilder._isValue(sets[property])) {
                continue;
            }
            const value = columns[property];
            if (!this.__sqlBuilder._isValue(value)) {
                continue;
            }
            sets[property] = value;
        }
        return this;
    }
    setIfHasNoValue(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            if (this.__sqlBuilder._isValue(sets[property])) {
                continue;
            }
            const value = columns[property];
            sets[property] = value;
        }
        return this;
    }
    setIfHasNoValueIfValue(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
            this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            if (this.__sqlBuilder._isValue(sets[property])) {
                continue;
            }
            const value = columns[property];
            if (!this.__sqlBuilder._isValue(value)) {
                continue;
            }
            sets[property] = value;
        }
        return this;
    }
    ignoreIfHasValue(...columns) {
        this.__query = '';
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
        }
        else {
            sets = this.__sets;
        }
        for (let i = 0, length = columns.length; i < length; i++) {
            let column = columns[i];
            if (!this.__sqlBuilder._isValue(sets[column])) {
                continue;
            }
            delete sets[column];
        }
        return this;
    }
    ignoreIfHasNoValue(...columns) {
        this.__query = '';
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
        }
        else {
            sets = this.__sets;
        }
        for (let i = 0, length = columns.length; i < length; i++) {
            let column = columns[i];
            if (this.__sqlBuilder._isValue(sets[column])) {
                continue;
            }
            delete sets[column];
        }
        return this;
    }
    ignoreAnySetWithNoValue() {
        this.__query = '';
        let sets;
        if (this.__onConflictUpdateSets) {
            sets = this.__onConflictUpdateSets;
        }
        else {
            sets = this.__sets;
        }
        const properties = Object.getOwnPropertyNames(sets);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            if (this.__sqlBuilder._isValue(sets[property])) {
                continue;
            }
            delete sets[property];
        }
        return this;
    }
    values(columns) {
        if (Array.isArray(columns)) {
            this.__isMultiple = true;
            if (columns.length == 1) {
                return this.set(columns[0]);
            }
            this.__multiple = columns;
            return this;
        }
        else {
            return this.set(columns);
        }
    }
    from(select) {
        this.__from = select;
        (0, ITableOrView_2.__addWiths)(select, this.__sqlBuilder, this.__withs);
        return this;
    }
    customizeQuery(customization) {
        this.__customization = customization;
        (0, ITableOrView_2.__addWiths)(customization.afterInsertKeyword, this.__sqlBuilder, this.__withs);
        (0, ITableOrView_2.__addWiths)(customization.afterQuery, this.__sqlBuilder, this.__withs);
        return this;
    }
    defaultValues() {
        this.__query = '';
        this.__sets = DEFAULT_VALUES;
        return this;
    }
    returningLastInsertedId() {
        this.__query = '';
        const table = this.__table;
        for (var columnName in table) {
            const column = (0, Column_1.__getColumnOfObject)(table, columnName);
            if (!column) {
                continue;
            }
            const columnPrivate = (0, Column_1.__getColumnPrivate)(column);
            if (!columnPrivate.__isAutogeneratedPrimaryKey) {
                continue;
            }
            if (this.__idColumn) {
                throw new Error('In order to call executeInsertReturningLastInsertedId method the table must have defined only one autogenerated primary key column');
            }
            this.__idColumn = column;
        }
        if (!this.__idColumn) {
            throw new Error('In order to call executeInsertReturningLastInsertedId method the table must have defined one autogenerated primary key column');
        }
        return this;
    }
    returning(columns) {
        this.__query = '';
        this.__columns = columns;
        this.__registerTableOrViewWithOfColumns(columns, this.__withs);
        return this;
    }
    returningOneColumn(column) {
        this.__query = '';
        this.__oneColumn = true;
        this.__columns = { 'result': column };
        (0, values_1.__getValueSourcePrivate)(column).__addWiths(this.__sqlBuilder, this.__withs);
        return this;
    }
    onConflictDoNothing() {
        this.__query = '';
        this.__onConflictDoNothing = true;
        return this;
    }
    onConflictDoUpdateDynamicSet() {
        this.__query = '';
        if (this.__onConflictUpdateSets) {
            throw new Error('Illegal state');
        }
        this.__onConflictUpdateSets = {};
        return this;
    }
    onConflictDoUpdateSet(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        if (this.__onConflictUpdateSets) {
            throw new Error('Illegal state');
        }
        this.__onConflictUpdateSets = {};
        let sets = this.__onConflictUpdateSets;
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            const value = columns[property];
            sets[property] = value;
        }
        this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        return this;
    }
    onConflictDoUpdateSetIfValue(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        if (this.__onConflictUpdateSets) {
            throw new Error('Illegal state');
        }
        this.__onConflictUpdateSets = {};
        let sets = this.__onConflictUpdateSets;
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            const value = columns[property];
            if (!this.__sqlBuilder._isValue(value)) {
                continue;
            }
            sets[property] = value;
        }
        this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        return this;
    }
    onConflictOn(...columns) {
        this.__query = '';
        if (this.__onConflictOnColumns) {
            throw new Error('Illegal state');
        }
        this.__onConflictOnColumns = columns;
        for (let i = 0, length = columns.length; i < length; i++) {
            (0, values_1.__getValueSourcePrivate)(columns[i]).__addWiths(this.__sqlBuilder, this.__withs);
        }
        return this;
    }
    onConflictOnConstraint(constraint) {
        this.__query = '';
        if (this.__onConflictOnConstraint) {
            throw new Error('Illegal state');
        }
        this.__onConflictOnConstraint = constraint;
        (0, ITableOrView_2.__addWiths)(constraint, this.__sqlBuilder, this.__withs);
        return this;
    }
    doNothing() {
        this.__query = '';
        this.__onConflictDoNothing = true;
        return this;
    }
    doUpdateDynamicSet() {
        this.__query = '';
        if (this.__onConflictUpdateSets) {
            throw new Error('Illegal state');
        }
        this.__onConflictUpdateSets = {};
        return this;
    }
    doUpdateSet(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        if (this.__onConflictUpdateSets) {
            throw new Error('Illegal state');
        }
        this.__onConflictUpdateSets = {};
        let sets = this.__onConflictUpdateSets;
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            const value = columns[property];
            sets[property] = value;
        }
        this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        return this;
    }
    doUpdateSetIfValue(columns) {
        this.__query = '';
        if (!columns) {
            return this;
        }
        if (this.__onConflictUpdateSets) {
            throw new Error('Illegal state');
        }
        this.__onConflictUpdateSets = {};
        let sets = this.__onConflictUpdateSets;
        const properties = Object.getOwnPropertyNames(columns);
        for (let i = 0, length = properties.length; i < length; i++) {
            const property = properties[i];
            const value = columns[property];
            if (!this.__sqlBuilder._isValue(value)) {
                continue;
            }
            sets[property] = value;
        }
        this.__valuesForInsert = this.__valuesForInsert || this.__getValuesForInsertOfColumns(columns);
        return this;
    }
    dynamicWhere() {
        this.__query = '';
        return this;
    }
    where(condition) {
        this.__query = '';
        if (this.__onConflictUpdateSets) {
            if (this.__onConflictUpdateWhere) {
                throw new Error('Illegal state');
            }
            this.__onConflictUpdateWhere = (0, values_1.asAlwaysIfValueSource)(condition);
            const conditionPrivate = (0, values_1.__getValueSourcePrivate)(condition);
            conditionPrivate.__addWiths(this.__sqlBuilder, this.__withs);
            this.__valuesForInsert = this.__valuesForInsert || conditionPrivate.__getValuesForInsert(this.__sqlBuilder);
        }
        else if (this.__onConflictOnColumns) {
            if (this.__onConflictOnColumnsWhere) {
                throw new Error('Illegal state');
            }
            this.__onConflictOnColumnsWhere = (0, values_1.asAlwaysIfValueSource)(condition);
            (0, values_1.__getValueSourcePrivate)(condition).__addWiths(this.__sqlBuilder, this.__withs);
        }
        else {
            throw new Error('Illegal state');
        }
        return this;
    }
    and(condition) {
        this.__query = '';
        if (this.__onConflictUpdateSets) {
            if (!this.__onConflictUpdateWhere) {
                this.__onConflictUpdateWhere = (0, values_1.asAlwaysIfValueSource)(condition);
            }
            else {
                this.__onConflictUpdateWhere = this.__onConflictUpdateWhere.and((0, values_1.asAlwaysIfValueSource)(condition));
            }
            const conditionPrivate = (0, values_1.__getValueSourcePrivate)(condition);
            conditionPrivate.__addWiths(this.__sqlBuilder, this.__withs);
            this.__valuesForInsert = this.__valuesForInsert || conditionPrivate.__getValuesForInsert(this.__sqlBuilder);
        }
        else if (this.__onConflictOnColumns) {
            if (!this.__onConflictOnColumnsWhere) {
                this.__onConflictOnColumnsWhere = (0, values_1.asAlwaysIfValueSource)(condition);
            }
            else {
                this.__onConflictOnColumnsWhere = this.__onConflictOnColumnsWhere.and((0, values_1.asAlwaysIfValueSource)(condition));
            }
            (0, values_1.__getValueSourcePrivate)(condition).__addWiths(this.__sqlBuilder, this.__withs);
        }
        else {
            throw new Error('Illegal state');
        }
        return this;
    }
    or(condition) {
        this.__query = '';
        if (this.__onConflictUpdateSets) {
            if (!this.__onConflictUpdateWhere) {
                this.__onConflictUpdateWhere = (0, values_1.asAlwaysIfValueSource)(condition);
            }
            else {
                this.__onConflictUpdateWhere = this.__onConflictUpdateWhere.or((0, values_1.asAlwaysIfValueSource)(condition));
            }
            const conditionPrivate = (0, values_1.__getValueSourcePrivate)(condition);
            conditionPrivate.__addWiths(this.__sqlBuilder, this.__withs);
            this.__valuesForInsert = this.__valuesForInsert || conditionPrivate.__getValuesForInsert(this.__sqlBuilder);
        }
        else if (this.__onConflictOnColumns) {
            if (!this.__onConflictOnColumnsWhere) {
                this.__onConflictOnColumnsWhere = (0, values_1.asAlwaysIfValueSource)(condition);
            }
            else {
                this.__onConflictOnColumnsWhere = this.__onConflictOnColumnsWhere.or((0, values_1.asAlwaysIfValueSource)(condition));
            }
            (0, values_1.__getValueSourcePrivate)(condition).__addWiths(this.__sqlBuilder, this.__withs);
        }
        else {
            throw new Error('Illegal state');
        }
        return this;
    }
    __addWiths(sqlBuilder, withs) {
        const withViews = this.__withs;
        for (let i = 0, length = withViews.length; i < length; i++) {
            const withView = withViews[i];
            (0, ITableOrView_1.__getTableOrViewPrivate)(withView).__addWiths(sqlBuilder, withs);
        }
    }
    __registerTableOrView(_sqlBuilder, _requiredTablesOrViews) {
        // do nothing because it is not possible to add external dependency
    }
    __registerRequiredColumn(_sqlBuilder, _requiredColumns, _onlyForTablesOrViews) {
        // do nothing because it is not possible to add external dependency
    }
    __getOldValues(_sqlBuilder) {
        // old values fake table is not possible to be used here
        return undefined;
    }
    __getValuesForInsert(_sqlBuilder) {
        // values for insert fake table is not possible to be used here
        return undefined;
    }
    __isAllowed(sqlBuilder) {
        let result = (0, ITableOrView_1.__getTableOrViewPrivate)(this.__table).__isAllowed(sqlBuilder);
        if (!result) {
            return false;
        }
        result = (0, ITableOrView_1.__isAllowed)(this.__from, sqlBuilder);
        if (!result) {
            return false;
        }
        const sets = this.__sets;
        for (let prop in sets) {
            const set = sets[prop];
            const result = (0, ITableOrView_1.__isAllowed)(set, sqlBuilder);
            if (!result) {
                return false;
            }
        }
        const multiple = this.__multiple;
        if (multiple) {
            for (let i = 0, length = multiple.length; i < length; i++) {
                const sets = multiple[i];
                for (let prop in sets) {
                    const set = sets[prop];
                    const result = (0, ITableOrView_1.__isAllowed)(set, sqlBuilder);
                    if (!result) {
                        return false;
                    }
                }
            }
        }
        if (this.__columns) {
            result = (0, SqlBuilder_1.isAllowedQueryColumns)(this.__columns, sqlBuilder);
            if (!result) {
                return false;
            }
        }
        result = (0, ITableOrView_1.__isAllowed)(this.__onConflictOnConstraint, sqlBuilder);
        if (!result) {
            return false;
        }
        result = (0, ITableOrView_1.__isAllowed)(this.__onConflictOnColumns, sqlBuilder);
        if (!result) {
            return false;
        }
        result = (0, ITableOrView_1.__isAllowed)(this.__onConflictOnColumnsWhere, sqlBuilder);
        if (!result) {
            return false;
        }
        const updateSets = this.__sets;
        for (let prop in updateSets) {
            const set = updateSets[prop];
            const result = (0, ITableOrView_1.__isAllowed)(set, sqlBuilder);
            if (!result) {
                return false;
            }
        }
        result = (0, ITableOrView_1.__isAllowed)(this.__onConflictUpdateWhere, sqlBuilder);
        if (!result) {
            return false;
        }
        if (this.__customization) {
            result = (0, ITableOrView_1.__isAllowed)(this.__customization.afterInsertKeyword, sqlBuilder);
            if (!result) {
                return false;
            }
            result = (0, ITableOrView_1.__isAllowed)(this.__customization.afterQuery, sqlBuilder);
            if (!result) {
                return false;
            }
        }
        return true;
    }
}
exports.InsertQueryBuilder = InsertQueryBuilder;
const DEFAULT_VALUES = {};
